docs/src/developer/typesystem.md

```diff
@@ -0,0 +1,217 @@
+must form a 'tree' in Nemo and AbstractAlgebra.
+
+One such possibility is union types. If a function should accept one of a
+number of concrete or abstract types that can't all be made to belong to a
+single abstract type due to this limitation then one can use a union type.
+
+For example, Nemo defines `RingElement` to be a union of `RingElem` and all
+the Julia standard types which behave like ring elements, e.g. all `Integer`
+types and types of rationals with `Integer` components.
+
+A second feature we make use of in Nemo is parameterised types. Each concrete
+and abstract type can take one or more parameters. These parameter can be any
+other type, either concrete or abstract. For example, in Julia `Rational{T}`
+is for rationals with numerator and denominator of type `T`.
+
+A great deal of control over parameterised types is possible, e.g. one can
+restrict the type parameter `T` using a `where` clause, e.g. to write a
+function that accepts all rational types with integer components of the same
+type one can use the type `Rational{T} where T <: Integer`.
+
+Nemo makes use of such parameterised types for generic ring constructions
+such as generic polynomial rings and matrices over a given base ring. The type
+of the elements of the base ring is substituted for the parameter `T` in any
+concrete instantiation of the types `Poly{T}` and `Mat{T}`, which are defined
+in AbstractAlgebra in `src/generic/GenericTypes.jl`.
+
+The totality of all univariate polynomial types, including those of generic
+`Poly{T}` types and those coming from C libraries (such as `fmpz_poly`), is
+represented by the abstract type `PolyElem{T}` which in turn belongs to
+`RingElem`, both defined in AbstractAlgebra in `src/AbstractTypes.jl`.
+
+Similarly, the totality of all matrix types, including explicit C types
+like `fmpz_mat` and the generic `Mat{T}` types is given by the abstract type
+`MatElem{T}`, again defined in AbstractAlgebra in `src/AbstractTypes.jl`.
+
+This hierarchy of types allows one to write functions at any level, e.g. for
+all univariate polynomial types, just those with a given base type `T`, or
+for a specific concrete type corresponding to just one kind of univariate
+polynomial.
+
+A third possibility to get around the single inheritance limitation of Julia is
+type traits. There is currently no explicit compiler/language support for
+traits, however various implementations exist that make use of type parameters
+in tricky ways. This allows one to add 'traits' to types, so long as those
+traits can be expressed as types. In this way, types can have multiple
+'properties' at the same time, instead of belonging to just a single abstract
+type.
+
+Nemo does not currently use type traits, though the map types in Nemo do make
+use of a custom analogue of this.
+
+Note that unlike class based systems that dispatch on the type of the implicit
+`this` parameter, Julia methods dispatch on the type of all arguments. This is
```

Comment:
Note that in some languages like Python, the dispatched-on parameter is not implicit and not necessarily called `this` (conventionally `self` in Python).

Reply:
Thanks. Fixed.